tests/tests/all/service_probe.rs

```diff
@@ -248,15 +250,75 @@ async fn builder_and_resolve_shall_succeed_when_ips_are_returned() {
         }
     }
 
-    assert!(
-        LoadBalancedChannel::builder(ServiceDefinition::from_parts("test.com", 5000).unwrap(),)
-            .lookup_service(SucceedResolve)
-            .timeout(tokio::time::Duration::from_millis(500))
-            .resolution_strategy(ginepro::ResolutionStrategy::Eager {
-                timeout: Duration::from_secs(20),
-            })
-            .channel()
+    LoadBalancedChannel::builder(ServiceDefinition::from_parts("test.com", 5000).unwrap())
+        .timeout(tokio::time::Duration::from_millis(500))
+        .resolution_strategy(ginepro::ResolutionStrategy::Eager {
+            timeout: Duration::from_secs(20),
+        })
+        .channel()
+        .await
+        .unwrap();
+}
+
+#[tokio::test]
+async fn builder_with_middleware_layers() {
+    let uris = Arc::new(std::sync::Mutex::new(Vec::new()));
+    let uris2 = Arc::clone(&uris);
+
+    let mut resolver = TestDnsResolver::default();
+    let probe_interval = tokio::time::Duration::from_millis(3);
+
+    let load_balanced_channel = LoadBalancedChannel::builder(("www.test.com", 5000))
+        .lookup_service(resolver.clone())
+        .dns_probe_interval(probe_interval)
+        .with_endpoint_layer(|endpoint: Endpoint| Some(endpoint.concurrency_limit(1)))
+        .with_endpoint_layer(move |endpoint: Endpoint| {
+            // record the uri so we can assert that all the layers are run
+            uris2.lock().unwrap().push(endpoint.uri().clone());
+            Some(endpoint)
+        })
+        .with_endpoint_layer(|endpoint: Endpoint| endpoint.user_agent("my ginepro client").ok())
```

Comment:
Couldn't this cause a silent bug in the sense that the function fails but you still build the channel and your configuration is not applied as a consequence. My question is: do we want to the callback to return the error so the internals can fail the operation?

Reply:
It does! The function returns an `Option<Endpoint>`, returning `None` will short circuit, and if it can't build the endpoint, it won't be added to list of endpoints

Reply:
This functions the same as if you got your TLS config wrong. That does report a warning though. I'll see if I can make these convenient fn impls auto emit warnings

Reply:
Yeah the awkward thing is that this is only the probe service which runs in a background task. If it would error and close the probe loop, it would not effect the channel in any way. 

I wonder if we can force the balanced channel to close in the event that the probe task closes

Reply:
What about making a wrapper type? only exposing the infallible methods? because `Endpoint` also has connect which we don't want anyone to call, and the tls config we already do internally in the lib. WDYT?

Reply:
@tl-helge-hoff are you asking for eagerly evaluated `Endpoint`s or am I misinterpreting?